### DISCUSSION TITLE
Unnecessary full compilations are performed when synchronizing BSP projects

### DISCUSSION THREAD

**Discussion by atty303** [com-lihaoyi/mill#1497] [Development]
When the BSP's `buildTargetScalacOptions` is requested, the following code will be executed.

https://github.com/com-lihaoyi/mill/blob/5266b791f33a9dea4e031a1a946d9bb080cd69ce/bsp/src/mill/bsp/MillBuildServer.scala#L448-L450

The `compileClasspath` target depends on the `localClasspath` target.

https://github.com/com-lihaoyi/mill/blob/5266b791f33a9dea4e031a1a946d9bb080cd69ce/scalalib/src/JavaModule.scala#L269-L275

Since the `localClasspath` target depends on the `compile` target, the compilation of all modules will be executed.
For BSP synchronization, I think it is enough to get the `dest` of the `compile` target, and there is no need to actually compile it.

I came up with the following potential fixes.

- Get only the `dest` of the `compile` target from the `localClasspath` target. (Seems impossible).
- Evaluate each element of the `compileClasspath` from the BSP separately, and do not evaluate the `compile`. However, this cannot handle the case where `compileClasspath` is customized.
- Skip compilation when evaluating targets from BSP (by communicate with ZincWorker)

Does anyone have any ideas?

**Comment by lefou** [comment]
I've come with a solution for the compilation problem in pull request: https://github.com/com-lihaoyi/mill/pull/1536
